sbt-bridge/src/dotty/tools/xsbt/CompilerBridgeDriver.java

```diff
@@ -50,15 +54,60 @@ public boolean sourcesRequired() {
     return false;
   }
 
+  private static VirtualFile asVirtualFileOrNull(SourceFile sourceFile) {
+    if (sourceFile.file() instanceof AbstractZincFile) {
+      return ((AbstractZincFile) sourceFile.file()).underlying();
+    } else {
+      return null;
```

Comment:
Instead of returning null and then throwing an exception, could we construct a ZincPlainFile or ZincVirtualFile on the fly based on sourceFile.file()? I wonder if we make this an instance method we could even update a cache at the same time.

Reply:
My thought here is that would be unexpected - shouldn't it be the case that the only thing we receive is what zinc gave us originally?

Reply:
@eed3si9n is there any rule about what VirtualFileRefs can be sent back to AnalysisCallback?

Reply:
If Zinc has a VirtualFile corresponding to our file we should use it, but if it doesn't (e.g., -sourcepath is being passed something which isn't in the list of input files), then I think it's fine to make up our own VirtualFile, that might break the machine-independence of the zinc cache but that's not a big problem compared to completely breaking compilation.

Reply:
My vague recollection is a loop that looks like this:

```
          xbti.VirtualFile        xbti.VirtualFile
┌────────┐          ┌────────┐    ┌───────────┐
│  sbt   ├─────────►│ Zinc   ├───►│Analysis   │
└────────┘          └─┬──────┘    └─▲─────────┘
                      │             │
           VirtualFile│             │
                   ┌──▼────────┐    │
                   │scalac/dotc│    │
                   └──┬────────┘    │
                      │             │
           VirtualFile│             │
                   ┌──▼─────────────┤
                   │AnalysisCallback│
                   └────────────────┘
```

Reply:
@smarter I now add a fallback file and not throw - with a warning that its probably a bug

Reply:
the warning is adequate because the fallback might not have the same `id` as the VirtualFile provided by Zinc, so incremental compilation becomes completely broken.